File: editplaylist.pl
#!/usr/bin/perl
# oyster - a perl-based jukebox and web-frontend
#
# Copyright (C) 2004 Benjamin Hanzelmann <ben@nabcos.de>,
#  Stephan Windmüller <windy@white-hawk.de>,
#  Stefan Naujokat <git@ethric.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

use CGI qw/:standard -no_xhtml/;
use URI::Escape;
use strict;
use oyster::conf;
use oyster::common;
use File::Find;

my %config = oyster::conf->get_config('oyster.conf');

my $frames = 1;
my $framestr = '';

if ((param('frames') && (param('frames') eq 'no'))) {
    $frames = 0;
    $framestr = '&frames=no';
}

if ($frames) {
    oyster::common->navigation_header();
  } else {
      oyster::common->noframe_navigation();
      print h1('Edit Playlist');
  }

my $cssdirclass = 'dir2';
my $cssfileclass = 'file2';
my $action = param('action') || '';
my $playlist = param('playlist') || '';
my @playlist = ();

if (($action eq 'edit') || ($action eq 'deletefile') || ($action eq 'deletedir')) {
    my $delfile = param('file') || '';
    my $deldir = param('dir') || '';

    print h1(a({href=>"editplaylist.pl?action=edit&playlist=${playlist}${framestr}"},$playlist));
    print a({href=>"editplaylist.pl?action=addbrowse&playlist=$playlist&dir=/${framestr}"},
	    'Add files to this list...'),br,br;

    # Get all entries from playlist and filter

    open (PLAYLIST, "$config{savedir}lists/$playlist") or die $!;
    while (my $line = <PLAYLIST>) {
	chomp($line);
	$line =~ s/^\Q$config{mediadir}\E/\//;
	if ($action eq 'edit' || (($action eq 'deletefile') && ($line ne $delfile))
	    || (($action eq 'deletedir') && !($line =~ /^\Q$deldir\E\/[^\/]*/))) {
	    push (@playlist, $line);
	}
    }

    @playlist = sort @playlist;

    if (($action eq 'deletefile') || ($action eq 'deletedir')) {

	# Write new playlist

	open (PLAYLIST, ">$config{savedir}lists/$playlist") || error_msg();

	sub error_msg {
	    print strong('Error: Could not save playlist!');
	    print end_html;
	    exit 0;
	}

	foreach my $line (@playlist) {
	    my $printline = $line;
	    $printline =~ s/^\///;
	    $printline = $config{mediadir} . $printline;
	    print PLAYLIST $printline . "\n";
	}
	close PLAYLIST;

    }


    listdir('/',0);

} elsif ($action eq 'addbrowse') {

    browse();

} elsif (($action eq 'adddir') || ($action eq 'addfile')) {
    my %filelist = ();
    my $toadd = param('toadd') || '';


    open (PLAYLIST, "$config{savedir}lists/$playlist") or die $!;
    while (my $line = <PLAYLIST>) {
	$filelist{$line} = 1;
    }

    $toadd =~ s/^\///;

    if ($action eq 'adddir') {
	find( { wanted => \&is_audio_file, no_chdir => 1 },
	      "$config{mediadir}$toadd");

	sub is_audio_file {
	    if ( ($_ =~ /ogg$/i) or ($_ =~ /mp3$/i) ) {
		$filelist{$_ . "\n"} = 1;
	    }
	}
    } elsif ($action eq 'addfile') {
	$filelist{$config{mediadir} . $toadd . "\n"} = 1;
    }
	
    open (FILELIST, ">$config{savedir}lists/$playlist") || error_msg();

	sub error_msg {
	    print strong('Error: Could not save playlist!');
	    print end_html;
	    exit 0;
	}

    foreach my $key (sort (keys %filelist)) {
	print FILELIST $key;
    }
    close(FILELIST);

    browse();

}

sub listdir {

    # listdir shows files from @playlist, sorted by directories
    # $basepath is cut away for recursive use

    my $basepath = $_[0];
    my $counter = $_[1];

    while (($counter < @playlist) && ($playlist[$counter] =~ /^\Q$basepath\E/)) {
	my $newpath = $playlist[$counter];
	$newpath =~ s/^\Q$basepath\E//;
	if ($newpath =~ /\//) {

	    # $newpath is directory and becomes the top one

	    $newpath =~ /^([^\/]*\/)/;
	    $newpath = $1;

	    # do not add padding for the top level directory

	    my $cutnewpath = $newpath;
	    my $escapeddir = '';
	    $cutnewpath =~ s/\/$//;
	    $cutnewpath = escapeHTML($cutnewpath);

	    if ($cssdirclass eq 'dir') {
		$cssdirclass = 'dir2';
	    } else {
		$cssdirclass = 'dir';
	    }

	    if (!($basepath eq '/')) {
		print "<div style='padding-left: 1em;'>";
		print "<table width='100%'><tr><td>";
		$escapeddir = uri_escape("$basepath$cutnewpath", "^A-Za-z");
		print b({-class=>$cssdirclass}, $cutnewpath);
		$newpath = "$basepath$newpath";
	    }  else {
		print "<table width='100%'><tr><td>";
		$escapeddir = uri_escape("/$cutnewpath", "^A-Za-z");
		print b({-class=>$cssdirclass}, $cutnewpath);
		$newpath = "/$newpath";
	    }

	    print "</td><td align='right'>";
	    print a({
		class=>$cssdirclass,
		href=>"editplaylist.pl?action=deletedir&dir=$escapeddir&" .
		    "playlist=${playlist}${framestr}"
		}, 'Delete');

	    print "</td></tr></table>";

            # Call listdir recursive, then quit padding with <div>

	    $counter = listdir($newpath,$counter);
	    if (!($basepath eq '/')) {
		print "</div>\n";
	    }
	} else {

	    # $newpath is a regular file without leading directory

	    print "<div style='padding-left: 1em;'>";
	    while ($playlist[$counter] =~ /^\Q$basepath\E/) {

		# Print all filenames in $basedir

		my $filename = $playlist[$counter];
		$filename =~ s/^.*\///;
		$filename =~ /(.*)\.(...)$/;
		my $nameonly = $1;
		my $escapedfile = uri_escape("$basepath$filename", "^A-Za-z");

		# $cssfileclass changes to give each other file
		# another color

		if ($cssfileclass eq 'file') {
		    $cssfileclass = 'file2';
		} else {
		    $cssfileclass = 'file';
		}
		print "<table width='100%'><tr>";
		print "<td align='left'>";

		print a({href=>"fileinfo.pl?file=${escapedfile}${framestr}",
			 class=>$cssfileclass}, $nameonly);

		print "</td>";

		print "<td align='right'>";

		print a({href=>"editplaylist.pl?action=deletefile" .
		    "&file=$escapedfile&playlist=${playlist}${framestr}",
		    class=>$cssfileclass}, 'Delete');

		print "</td>";

		print "</tr></table>\n";
		$counter++;
	    }
	    print "</div>\n";
	}
    }

    return ($counter);

}

sub browse {

    print h1(a({href=>"editplaylist.pl?action=edit&playlist=${playlist}${framestr}"},$playlist));

    my $givendir = '/';

    if (param('dir')) {
	$givendir=param('dir') . "/";
	$givendir =~ s@//$@/@;
	$givendir =~ s/\.\.\///g;
	$givendir = '/' if ($givendir eq "..");
    }

    if ((!($givendir eq '/')) && (-e "$config{mediadir}$givendir")) {

	print "<p><strong>Current directory: ";

	my @dirs = split(/\//, $givendir);
	my $incdir = '';
	foreach my $partdir (@dirs) {
	    my $escapeddir = uri_escape("$incdir$partdir", "^A-Za-z");
	    my $escapedpartdir = $partdir;
	    $escapedpartdir =~ s/&/&amp;/g;
	    print a({href=>"editplaylist.pl?action=addbrowse&playlist=$playlist" .
		"&dir=${escapeddir}${framestr}"}, $escapedpartdir);

	    print "/ ";
	    $incdir = $incdir . "$partdir/";
	}

	print "</strong></p>";

	my $topdir = $givendir;
	$topdir =~ s/\Q$config{mediadir}\E//;
	if ($topdir =~ /^[^\/]*\/$/) {
	    $topdir = '';
	} else {
	    $topdir =~ s/\/[^\/]*\/$//;
	}

	my $escapeddir = uri_escape($topdir, "^A-Za-z");
	print a({href=>"editplaylist.pl?action=addbrowse" .
	    "&playlist=$playlist&dir=${escapeddir}${framestr}"}, 'One level up'), br, br;
	
    } elsif (!(-e "$config{mediadir}$givendir")) {   
	print h1('Error!');
	print "The directory $givendir could not be found.";
	print end_html;
    }

    my $globdir = "$config{mediadir}$givendir";
    $globdir =~ s/\ /\\\ /g;
    my @entries = <$globdir*>;

    print "<table width='100%'>";
    
    my (@files, @dirs) = ();
    
    foreach my $entry (@entries) {
	if (-d "$entry") {
	    push (@dirs, "$entry");
	} elsif (-f "$entry") {
	    push (@files, "$entry");
	}
    }

    my $cssdirclass = 'dir2';
    my $cssfileclass = 'file2';
    my $csslistclass = 'playlist2';
    my $anchorcounter = 0;

    foreach my $dir (@dirs) {
	if ($cssdirclass eq 'dir') {
	    $cssdirclass = 'dir2';
	} else {
	    $cssdirclass = 'dir';
	}

	$dir =~ s/\Q$config{mediadir}\E//;
	my $escapeddir = uri_escape("$dir", "^A-Za-z");
	$dir =~ s/^.*\///;
	$dir =~ s/&/&amp;/g;

	print "<tr>";
	print "<td><a name='a" . $anchorcounter . "'></a>";
	print a({class=>$cssdirclass,
		 href=>"editplaylist.pl?action=addbrowse&playlist=$playlist" .
		     "&dir=${escapeddir}${framestr}"}, $dir);
	print "</td>";

	print "<td align='right'>";
	print a({class=>$cssdirclass,
		href=>"editplaylist.pl?action=adddir&playlist=$playlist" .
		"&toadd=$escapeddir&dir=${givendir}${framestr}#a" . $anchorcounter++}, 'Add');

	print "</td></tr>\n";
    }
    
    foreach my $file (@files) {
	$file =~ s/\Q$config{mediadir}$givendir\E//;
	print "<tr>";
	if (($file =~ /mp3$/i) || ($file =~ /ogg$/i)) {
	    my $escapeddir = "$givendir$file";
	    $escapeddir =~ s/\Q$config{mediadir}\E//;
	    $escapeddir = uri_escape("$escapeddir", "^A-Za-z");
	    if ($cssfileclass eq 'file') {
		$cssfileclass = 'file2';
	    } else {
		$cssfileclass = 'file';
	    }
	    my $escapedfile = $file;
	    $escapedfile =~ s/&/&amp;/g;
	    print "<td><a name='a" . $anchorcounter . "'></a>";

	    print a({class=>$cssfileclass,
		     href="fileinfo.pl?file=${escapeddir}${framestr}"}, $escapedfile);
	    print "</td>";

	    print "<td align='right'>";

	    print a({class=>$cssfileclass,
		     href="editplaylist.pl?action=addfile&playlist=$playlist" .
			 "&toadd=$escapeddir&dir=${givendir}${framestr}#a" . $anchorcounter++}, 'Add');

	}
	print "</td></tr>\n";
    }

    print "</table>";
    print end_html;


}

File: editplaylist.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

# oyster - a python-based jukebox and web-frontend
#
# Copyright (C) 2004 Benjamin Hanzelmann <ben@nabcos.de>,
#  Stephan Windmüller <windy@white-hawk.de>,
#  Stefan Naujokat <git@ethric.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

import cgi
import common
import config
import fifocontrol
import cgitb
import sys
import os.path
import urllib
cgitb.enable()


def print_frameset():
    
    """Generates a frameset for the playlist editor"""
    
    print "Content-Type: text/html; charset=" + myconfig['encoding'] + "\n"
    print """
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN"
   "http://www.w3.org/TR/html4/frameset.dtd">
<html>
<head>
"""
    print"<title>Edit Playlist " + playlist + "</title>"
    print"<link rel='shortcut icon' href='themes/default/favicon.png'>"
    print"</head>"
    print"<frameset rows='50,*'>"
    print"  <frame src='editplaylist.py?mode=title&playlist=" + urllib.quote(playlist) + "' name='title'>"
    print"  <frameset cols='*,*'>"
    print"   <frame src='editplaylist.py?mode=edit&playlist=" + urllib.quote(playlist) + "' name='playlist'>"
    print"   <frame src='browse.py?mode=editplaylist&playlist=" + urllib.quote(playlist) + "' name='browse'>"
    print"  </frameset>"
    print"    <noframes>"
    print"	<p>"
    print"   Your browser does not seem to support display of frames."
    print"	</p>"
    print"  </noframes>"
    print"</frameset>"
    print"</html>"


def print_title():
    print "Content-Type: text/html; charset=" + myconfig['encoding'] + "\n"
    print "<?xml version='1.0' encoding='" + myconfig['encoding'] + "' ?>"
    print """
    <!DOCTYPE html 
         PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
              "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
    <html xmlns="http://www.w3.org/1999/xhtml">
    <head>
     <title>Oyster-GUI</title>
    """ 
    print "<meta http-equiv='Content-Type' content='text/html; charset=" + myconfig['encoding'] + "' />"
    print "<link rel='stylesheet' type='text/css' href='themes/" + myconfig['theme'] + "/layout.css' />"
    print "<link rel='shortcut icon' href='themes/" + myconfig['theme'] + "/favicon.png' />"
    print "</head><body>"
    print "<form method='post' target='_top' action='index.html'><input type='hidden' name='view' value='playlists'>"
    print "<p align='center'><b>Editing playlist " + playlist + "</b> " +\
        "<input type='submit' value='Done'></p></form>"
    print "</body></html>"
    
myconfig = config.get_config()
basedir = myconfig['basedir']
savedir = myconfig['savedir']
mediadir = myconfig['mediadir'][:-1]
form = cgi.FieldStorage()

if 'action' in form and form['action'].value == 'addnewlist' and 'newlistname' in form:
    fifocontrol.do_action('addnewlist', form['newlistname'].value)
        
if 'playlist' in form:
    playlist = form['playlist'].value
elif 'newlistname' in form:
    playlist = form['newlistname'].value
else:
    common.navigation_header(title="Oyster-GUI")
    print "<p>You did not specify a name for the playlist.</p>"
    print "<p>Please press the <i>Back</i> button in your browser and try again.</a></p>"
    common.html_footer()
    sys.exit()

if playlist == 'default':
    common.navigation_header(title="Oyster-GUI")
    print "<p>It is not allowed to edit the default playlist.</p>"
    common.html_footer()
    sys.exit()

if not 'mode' in form and not 'delfile' in form and not 'deldir' in form and not 'addfile' in form\
        and not 'adddir' in form:
    print_frameset()
    sys.exit()
elif 'mode' in form and form['mode'].value == 'title':
    print_title()
    sys.exit()

# Starting from here: mode == edit

common.html_header(title="Oyster-GUI")

allfiles = []
playlistfile = open(savedir + "lists/" + playlist)
for line in playlistfile.readlines():
    line = line.replace(mediadir, '', 1)
    allfiles.append(line[:-1])
playlistfile.close()

# Delete a single file

if 'delfile' in form:
    allfiles.remove(form['delfile'].value)

# Delete a complete directory

if 'deldir' in form:
    tmpfiles = []
    for tmpfile in allfiles:
        if tmpfile.find(form['deldir'].value) != 0:
            tmpfiles.append(tmpfile)
    allfiles = tmpfiles

# Add a single file

if 'addfile' in form:
    if form['addfile'].value not in allfiles:
        allfiles.append(form['addfile'].value)

# Add a complete directory

filetypes = myconfig['filetypes'].lower().split(',')

if 'adddir' in form:
    for root, dirs, files in os.walk(mediadir + form['adddir'].value, topdown=False):
        for name in files:
            if name[name.rfind(".")+1:].lower() in filetypes:
                root = root.replace(mediadir, '', 1)
                if os.path.join(root, name).rstrip() not in allfiles:
                    allfiles.append(os.path.join(root, name).rstrip())
        

allfiles.sort()

if 'addfile' in form or 'adddir' in form or 'delfile' in form or 'deldir' in form:
    playlistfile = open(savedir + "lists/" + playlist, "w")
    for curfile in allfiles:
        playlistfile.write(mediadir + curfile + "\n")
    playlistfile.close()

import common
common.results = allfiles
allfiles = common.sort_results('/')

common.listdir('/', 0, 'file2', 1, playlist)

common.html_footer()
